### PR TITLE
scheduler: decreate the repeat filter in scheduler. 

### LINKED ISSUE #5539: some filter should run once
The storeStateFilter, SpecialUseFilter, RegionScoreFilter can be run once , not every target. 
some related code:
https://github.com/tikv/pd/blob/0433eea9bf6de3526fa40b8a96a88ac7891fc9b0/server/schedulers/balance_region.go#L230-L236

### PR DESCRIPTION
**Pull request by bufferflies**
### What problem does this PR solve?

Issue Number: Close #5539

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
- Integration test
image

Code changes

Side effects

Related changes

### Release note

```release-note
None.
```

### CODE REVIEW

File: server/schedulers/balance_region.go [hunk 1]
```diff
@@ -232,25 +235,27 @@ func (s *balanceRegionScheduler) Schedule(cluster schedule.Cluster, dryRun bool)
...
-		&filter.StoreStateFilter{ActionScope: s.GetName(), MoveRegion: true},
-		filter.NewRegionScoreFilter(s.GetName(), solver.source, solver.GetOpts()),
+		filter.NewExcludedFilter(s.GetName(), nil, excludeTargets),
 		filter.NewPlacementSafeguard(s.GetName(), solver.GetOpts(), solver.GetBasicCluster(), solver.GetRuleManager(), solver.region, solver.source),
 	}
 
-	candidates := filter.NewCandidates(solver.GetStores()).
-		FilterTarget(solver.GetOpts(), collector, filters...).
-		Sort(filter.RegionScoreComparer(solver.GetOpts()))
+	candidates := filter.NewCandidates(dstStores).FilterTarget(solver.GetOpts(), collector, filters...)
```
Comment (rleungx) [inline server/schedulers/balance_region.go:250]: Will it affect the current implementation of diagnostic?

File: server/schedulers/utils.go [hunk 2]
```diff
@@ -79,69 +80,87 @@ func (p *solver) TargetMetricLabel() string {
...
+			zap.Int64("tolerant-resource", p.getTolerantResource()))
 	}
 	return shouldBalance
 }
 
 func (p *solver) getTolerantResource() int64 {
-	if p.kind.Resource == core.LeaderKind && p.kind.Policy == core.ByCount {
-		return int64(p.tolerantSizeRatio)
+	if p.tolerantSource > 0 {
+		return p.tolerantSource
```
Comment (rleungx) [inline server/schedulers/utils.go:154]: What if the average size changes?

File: server/core/store.go [hunk 3]
```diff
@@ -209,6 +209,9 @@ func (s *StoreInfo) GetLabels() []*metapb.StoreLabel {
 
 // GetID returns the ID of the store.
 func (s *StoreInfo) GetID() uint64 {
+	if s == nil {
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply server/core/store.go]: fix

File: server/schedulers/balance_region.go (hunk 1 above)
Comment (bufferflies) [reply server/schedulers/balance_region.go:258]: yes, I will record the reason why the store can not be selected and put it into the collector.

**Review by CabinfeverB — COMMENTED** [review]
Rest LGTM

File: server/schedulers/balance_region.go [hunk 4]
```diff
@@ -171,8 +172,15 @@ func (s *balanceRegionScheduler) Schedule(cluster schedule.Cluster, dryRun bool)
 	}
 
 	solver.step++
-	for _, solver.source = range stores {
+	var sourceIndex int
+
+	// sourcesStore is sorted by region score desc, so we pick the first store as source store.
+	for sourceIndex, solver.source = range sourceStores {
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply server/schedulers/balance_region.go:178]: In this pr, the `region score filter ` has been removed, I think this status can be remove also.

File: server/schedulers/utils.go [hunk 5]
```diff
@@ -79,69 +80,87 @@ func (p *solver) TargetMetricLabel() string {
...
-		opInfluenceStatus.WithLabelValues(scheduleName, strconv.FormatUint(sourceID, 10), "source").Set(float64(sourceInfluence))
-		opInfluenceStatus.WithLabelValues(scheduleName, strconv.FormatUint(targetID, 10), "target").Set(float64(targetInfluence))
-		tolerantResourceStatus.WithLabelValues(scheduleName, strconv.FormatUint(sourceID, 10), strconv.FormatUint(targetID, 10)).Set(float64(tolerantResource))
+		targetDelta := influence*influenceAmp + tolerantResource
+		score = p.target.RegionScore(opts.GetRegionScoreFormulaVersion(), opts.GetHighSpaceRatio(), opts.GetLowSpaceRatio(), targetDelta)
 	}
+	return score
+}
+
+func (p *solver) shouldBalance(scheduleName string) bool {
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply server/schedulers/utils.go:133]: fix

**Comment by CabinfeverB** [comment]
I think the top score store will not have diagnostic info, I will test it.

File: server/schedulers/utils.go [hunk 6]
```diff
@@ -79,69 +80,89 @@ func (p *solver) TargetMetricLabel() string {
...
-		p.sourceScore = p.source.LeaderScore(p.kind.Policy, sourceDelta)
-		p.targetScore = p.target.LeaderScore(p.kind.Policy, targetDelta)
+		targetDelta := influence + tolerantResource
+		score = p.target.LeaderScore(p.kind.Policy, targetDelta)
 	case core.RegionKind:
-		sourceDelta, targetDelta := sourceInfluence*influenceAmp-tolerantResource, targetInfluence*influenceAmp+tolerantResource
-		p.sourceScore = p.source.RegionScore(opts.GetRegionScoreFormulaVersion(), opts.GetHighSpaceRatio(), opts.GetLowSpaceRatio(), sourceDelta)
-		p.targetScore = p.target.RegionScore(opts.GetRegionScoreFormulaVersion(), opts.GetHighSpaceRatio(), opts.GetLowSpaceRatio(), targetDelta)
-	}
-	if opts.IsDebugMetricsEnabled() {
```
Comment (rleungx) [inline server/schedulers/utils.go:115]: Is it deprecated?

File: server/schedulers/balance_test.go [hunk 7]
```diff
@@ -780,9 +787,9 @@ func TestBalanceRegionSchedule1(t *testing.T) {
 	// test region replicate not match
 	opt.SetMaxReplicas(3)
 	ops, plans := sb.Schedule(tc, true)
-	re.Len(plans, 100)
+	re.Greater(len(plans), 100)
 	re.Empty(ops)
-	re.Equal(int(plans[0].GetStatus().StatusCode), plan.StatusRegionNotReplicated)
+	re.Equal(int(plans[0].GetStatus().StatusCode), plan.StatusStoreRemoving)
```
[the remark this answers is not in the document]
Comment (CabinfeverB) [reply server/schedulers/balance_test.go:792]: The search space has changed

File: server/schedulers/utils.go (hunk 6 above)
Comment (bufferflies) [reply server/schedulers/utils.go:115]: No, I will add it again

File: server/schedulers/balance_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (rleungx) [reply server/schedulers/balance_test.go:792]: The comment is misleading.

File: server/schedule/filter/filters.go [hunk 8]
```diff
@@ -51,6 +51,29 @@ func SelectSourceStores(stores []*core.StoreInfo, filters []Filter, opt *config.
 	})
 }
 
+// SelectFaultTargetStores selects fault stores that can't be selected as target store from the list.
+func SelectFaultTargetStores(stores []*core.StoreInfo, filters []Filter, opt *config.PersistOptions, collector *plan.Collector) []*core.StoreInfo {
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply server/schedule/filter/filters.go]: fix

File: server/schedulers/balance_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (bufferflies) [reply server/schedulers/balance_test.go]: yes， I will keep the result.

**Review by CabinfeverB — APPROVED** [review]

**Pull request merged**